pageserver/src/tenant.rs

```diff
@@ -350,9 +361,604 @@ impl Drop for TimelineUninitMark {
+                let timeline_uninit_mark_file = self
+                    .conf
+                    .timeline_uninit_mark_file_path(self.tenant_id, timeline_id);
+                if timeline_uninit_mark_file.exists() {
+                    info!(
+                        "Found an uninit mark file for timeline {}/{}, removing the timeline and its uninit mark",
+                        self.tenant_id, timeline_id
+                    );
+                    if let Err(e) =
+                        remove_timeline_and_uninit_mark(&timeline_dir, &timeline_uninit_mark_file)
+                    {
+                        error!("Failed to clean up uninit marked timeline: {e:?}");
+                    }
+                    continue;
+                }
+
+                // TODO There was also remove if dir is empty logic
+                // TODO if attach failed we can be missing metadata here
+
+                let file_name = entry.file_name();
+                if let Ok(timeline_id) =
+                    file_name.to_str().unwrap_or_default().parse::<TimelineId>()
+                {
+                    let metadata = load_metadata(self.conf, timeline_id, self.tenant_id)
+                        .context("failed to load metadata")?;
+                    timelines_to_load.insert(timeline_id, metadata);
+                } else {
+                    // A file or directory that doesn't look like a timeline ID
+                    warn!(
+                        "unexpected file or directory in timelines directory: {}",
+                        file_name.to_string_lossy()
+                    );
+                }
+            }
+        }
+
+        // Sort the array of timeline IDs into tree-order, so that parent comes before
+        // all its children.
+        let sorted_timelines = tree_sort_timelines(timelines_to_load)?;
+        // FIXME if we init some of the timelines and one of them is broken we
+        //      potentially will end up with partially launched wal receivers.
+        //    1. Either move it out of initialize_with_lock
+        //    2. Or pass it as a flag
+        // FIXME original collect_timeline_files contained one more check:
+        //    1. "Timeline has no ancestor and no layer files"
+
+        // XXX get rid of enable_background_jobs
+        let enable_background_jobs = !sorted_timelines.is_empty();
+
+        for (timeline_id, local_metadata) in sorted_timelines {
+            // FIXME should we fail load of whole tenant if one timeline failed?
+            //   consider branch hierarchy. Maybe set one to broken and others to Paused or something
+            self.load_local_timeline(timeline_id, local_metadata)
+                .await?;
+        }
+
+        // We're ready for business.
+        // Change to active state under the hood spawns background loops
+        // The loops will shut themselves down when they notice that the tenant is inactive.
+        self.activate(enable_background_jobs);
```

Comment:
The concerns from `spawn_attach`'s call to this function would apply here as well...

Reply:
Addressed by https://github.com/neondatabase/neon/pull/2910